MSC on-board flash: Persist RTC across reboot and use as file timestamp

etracer65 opened the pull request:
NOTE: Requires development version of Configurator 10.5 for proper RTC clock operation. Changes were made for API version 1.41 that effect the setting of the RTC and it will not work properly with older versions of the Configurator (changes are unrelated to this PR).

Adds support to persist the RTC (if set) across the reboot if entering mass storage mode for on-board flash. The value is then used as the timestamp for the files exposed in the virtual FAT32 filesystem. The files will then have reasonable creation dates when copied to the host computer.

The RTC is set when the Configurator connects to the firmware. So the RTC value is available when the user reboots the flight controller in to mass storage mode using the `msc` command.

If the RTC is not set (or supported), then the default timestamp of 2018-01-01 will be used (unchanged from previous).

Also stubbed in support for `PERSISTENT` memory region on F7. So the feature will work once `PERSISTENT` support is fully added.

Included some improvements to the RTC functions and exposed the `tz_offset_minutes` in the `timeConfig` PG. Support already existed for timezone offsets but the parameter wasn't exposed to the user and couldn't be set.

etracer65 commented:
I built in support for F7 once `PERSISTENT` memory block support is added for the platform. So currently this PR is not functional on F7, but should have no negative side effects. Should be tested on F7 with on-board flash to confirm no negative impact (I don't have one).

mikeller commented on src/main/target/common_post.h:
```diff
@@ -214,3 +214,7 @@
 #endif
+
+#if (defined(USE_FLASHFS) && defined(USE_RTC_TIME) && defined(USE_USB_MSC))
+#define PERSIST_MSC_RTC
```
Can I suggest we do this the other way round? I.e. define `PERSIST_MSC_RTC` in `common_pre.h` for MCU types we want it in, and then undefine in `common_post.h` if any of the prerequisites are not defined? This way, individual targets can undefine it explicitly if there's a reason to do so.

etracer65 replied:
Makes perfect sense. Any preferences on where in `common_pre.h` the `#define PERSIST_MSC_RTC` lives? Could either be in the `#if (FLASH_SIZE > 256)` section or in the `#ifdef STM32F4` and `#ifdef STM32F7` sections. I guessing probably the latter as that's also where `USE_USB_MSC` is defined - both either way would work.  I guess I should also rename it if we're going this way to something like `USE_PERSISTENT_MSC_RTC`.

mikeller replied:
The F4 / F7 sections are probably the best - anything that is dependent on MCU type (and not just a question of firmware size) should go in there.

etracer65 replied:
Ammended as suggested. Renamed to `USE_PERSISTENT_MSC_RTC`.

mikeller commented on src/main/msc/emfat_file.c:
```diff
@@ -327,8 +337,16 @@ void emfat_init_files(void)
 
+#ifdef USE_PERSISTENT_MSC_RTC
+    if (mscRebootRtc > 0) { // Only supporting dates after the unix epoch start
+        const int32_t rtcSeconds = rtcTimeGetSeconds(&mscRebootRtc) + (timeConfig()->tz_offsetMinutes * 60);
```
Just tested this under linux, and the offset does not work as it should (local time when testing 20181202, 1155 NZDT):

![image](https://user-images.githubusercontent.com/4742747/49333854-5848e900-f62b-11e8-9983-4f42c098b3f5.png)

Setting `timezone_offset_minutes` to `0` will make the file timestamps to appear correct - could it be that they are meant to be UTC?

etracer65 replied:
I just retested on OS X, Windows 7 and Windows 10 and all show the correct local time for the file timestamps with my -300 offset.

But in these cases I set the time in the flight controller with an OS X build of the configurator and then connected the USB device to VM's running windows. Don't have a way to build Windows versions of the Configurator at this point. Maybe the timebase used to set the time varies by the platform the Configurator is running on? For OS X at least it sets the time as UTC.

mikeller replied:
If I set `timezone_offset_minutes = 0` I get UTC in CLI `status`. If I set `timezone_offset_minutes = 780` I get local time in CLI `status` (like on the screenshot above).

etracer65 replied:
Well it sounds like the time from the configurator is consistent as UTC. I installed node.js on a Windows 10 vm and build the configurator from current master and tested completely in Windows. The timestamps were all correct with my -300 offset (as is the time in `status`). Same for OS X.  So maybe the Linux FAT32 filesystem implementation is behaving differently.

etracer65 replied:
I did some research and FAT32 has no concept of timezones in the file timestamps. In NTFS the specification is they're stored as UTC, but not in FAT32.

mikeller replied:
Probably just something we have to live with on linux.

etracer65 replied:
Googling a little I’m getting the impression that the UTC assumption is intentional as apparently many cameras, etc. store the files as UTC. So it seems a bit of hack to deal with this.

I suppose we could always add some kind of “use UTC for MSC” flag to work around the workaround. Or maybe it’s not worth it for the limited Linux user base?

mikeller commented on src/main/interface/settings.c:
```diff
@@ -1243,6 +1243,11 @@ const clivalue_t valueTable[] = {
+
+// PG_TIMECONFIG
+#ifdef USE_RTC_TIME
+    { "timezone_offset_minutes",  VAR_INT16 | MASTER_VALUE, .config.minmax = { -1440, 1440 }, PG_TIME_CONFIG, offsetof(timeConfig_t, tz_offsetMinutes) },
```
I think the accepted / valid range for the offset is +/- 13 hours (780 min).

etracer65 commented:
I pushed an additional commit that abstracts the timezone offset up one layer to `systemResetToMsc()`.

Adds support for specifying a custom timezone offset from both the CLI and MSP calls to enter mass storage mode.

Added an option timezone offset minutes to the CLI `msc` command. If no parameter is specified then the default as specified by `timezone_offset_minutes` will be used. So to reboot into mass storage mode and force the file timestamps to be in UTC, use `msc 0`.

Support can be added to MSP method of invoking `systemResetToMsc()` - currently it defaults to the configured `timezone_offset_minutes`. As this will require discussion and integration with the Configurator I'll leave this to a separate PR.

mikeller commented on src/main/drivers/usb_msc_f4xx.c:
```diff
@@ -145,11 +145,20 @@ void mscWaitForButton(void)
     *((uint32_t *)0x2001FFF0) = MSC_MAGIC;
 
     __disable_irq();
+    
```
Nitpicking here, but shouldn't this be done before diabling IRQs?

etracer65 replied:
Not sure it matters as we're doing a reboot a couple of instructions later. But I'll have other changes so I can revise.

mikeller commented on src/main/common/time.c:
```diff
@@ -238,6 +238,17 @@ bool rtcGet(rtcTime_t *t)
+{
+    rtcTime_t workTime;
+    if (rtcGet(&workTime)) {
+        *t = workTime + (minutes * 60 * MILLIS_PER_SECOND);
```
This should check for underflow.

etracer65 replied:
The `rtcTime_t` type is `int64_t` so underflow really isn't a concern.

mikeller replied:
Interesting. I had thought it started at the beginning of the epoch.

etracer65 replied:
It is, but there were dates before 1970 ;). So the standard implementation (and how the RTC code is built) uses signed values.  I agree that in our context it really doesn't make sense. But keeping our library per the standard is probably a good idea in case we interact with other systems that speak "unix epoch time".

mikeller replied:
Agreed there. That was also the motivation for my suggestion to stick the full RTC timestamp into the permanent memory.

etracer65 replied:
This is such a limited use-case though that the loss of precision to seconds (or even minutes) is fine so no real reason to persist the entire millisecond-level 64bit timestamp.

When I started this I did investigate generally persisting the RTC time across reboots and resetting on initialization. But it would take hitting every possible occurrence of the reboot code and there are a lot. Without doing it deep in the STM libraries and changing `NVIC_SystemReset()` it would be a lot of work to pull off. And for probably questionable value as the RTC time would be behind anyway (due to the reboot/init process) and there are a limited number of use cases where having the RTC be restored after a reboot (assuming it was previously set) are of value. The main one being for MSC file timestamps so I went with a more vertical approach since we had a clear exit point for rebooting into MSC.

Also, dealing with actual persisted value is painful. We have to assume that there will be cases where the value won't be set (like a cold boot). So we have to presume a 0 value is unset. But a 0 value is technically valid timestamp. However we can ignore that in our limited use case as we're unlikely to have a 1970 date for the flight controller RTC and we can just consider that to be invalid. Becomes more challenging if considering a generalized RTC persistence.

mikeller commented:
Re MSP, should we add an `int16_t` parameter to `MSP_REBOOT_MSC` that allowed configurator to supply a custom offset, with `INT16_MAX` indicating 'use `timeConfig()->tz_offsetMinutes`'?

etracer65 commented:
Re. MSP - I thought the same. But it's a little messy how the reboot is handled in msp.c. It's passed through a dispatcher `mspRebootFn` which is passed the serial port and handles the different types of reboots. Would be messy to add the offset minutes to the parameters as they only affect one type. Probably just need to make the offset minutes a global even though that's a little unpleasant. The `rebootMode` is already working this way so I guess it's okay.

etracer65 commented:
Also #7166 (if it goes forward) will introduce substantial changes to this PR on how persistent values are stored/retrieved. So I'm a bit in a holding pattern until we sort that out.

mikeller commented:
While this is waiting for the persistence layer to change, how about adding this to MSP by adding a `MSP_REBOOT_MSC_UTC` value for `rebootMode`: https://github.com/betaflight/betaflight/blob/4442de55919f2001cbe6546c811a4481b2932e5b/src/main/interface/msp.c#L1421

This will give us the functionality we need to make it work seamlessly in linux, without forcing us to extend the command format.

etracer65 commented:
Rebased and squashed.

Revised to use the new persistence layer. Added support for a new MSP reboot type `MSP_REBOOT_MSC_UTC` that will override any user timezone offset with 0 for UTC time.
